content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,608 @@
+        label.
+
+    Returns:
+        A tuple (x, y). The element x is a tuple containing (1) a tensor of padded 
+        word vectors and (2) their respective lengths of the sequences. The element 
+        y is a tensor of padded tag indices. The word vectors are padded with vectors 
+        of 0s and the tag indices are padded with -100s. Padding with -100 is done 
+        because of the cross-entropy loss and the accuracy metric ignores 
+        the targets with values -100.
+    """
+
+    # This gets us two lists of tensors and a list of integer. 
+    # Each tensor in the first list is a sequence of word vectors.
+    # Each tensor in the second list is a sequence of tag indices.
+    # The list of integer consist of the lengths of the sequences in order.
+    sequences_vectors, sequences_labels, lengths = zip(*[
+        (torch.FloatTensor(seq_vectors), torch.LongTensor(labels), len(seq_vectors)) 
+         for (seq_vectors, labels) in sorted(batch, key=lambda x: len(x[0]), reverse=True)
+    ])
+
+    lengths = torch.LongTensor(lengths)
+
+    padded_sequences_vectors = pad_sequence(sequences_vectors, batch_first=True, padding_value=0)
+
+    padded_sequences_labels = pad_sequence(sequences_labels, batch_first=True, padding_value=-100)
+
+    return (padded_sequences_vectors, lengths), padded_sequences_labels
+```
+
+```python
+train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True, collate_fn=pad_collate_fn)
+valid_loader = DataLoader(valid_data, batch_size=batch_size, collate_fn=pad_collate_fn)
+test_loader = DataLoader(test_data, batch_size=batch_size, collate_fn=pad_collate_fn)
+```
+
+## Full Network
+> We use a second trick, ``packing``.
+
+Since our sequences are of variable lengths and that we want to be as efficient as possible when packing them, we cannot use the [PyTorch `nn.Sequential`](https://pytorch.org/docs/stable/generated/torch.nn.Sequential.html) class to define our model. Instead, we define the forward pass so that it uses packed sequences (again, you can read [this good explanation](https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch) of why we pack sequences).
+
+```python
+class FullNetWork(nn.Module):
+    def __init__(self, lstm_network, fully_connected_network):
+        super().__init__()
+        self.hidden_state = None
+        
+        self.lstm_network = lstm_network
+        self.fully_connected_network = fully_connected_network
+        
+    def forward(self, padded_sequences_vectors, lengths):
+        """
+            Defines the computation performed at every call.
+        """
+        total_length = padded_sequences_vectors.shape[1]
+        pack_padded_sequences_vectors = pack_padded_sequence(padded_sequences_vectors, lengths.cpu(), batch_first=True)
+
+        lstm_out, self.hidden_state = self.lstm_network(pack_padded_sequences_vectors)
+        lstm_out, _ = pad_packed_sequence(lstm_out, batch_first=True, total_length=total_length)
```

Comment:
Dimension comments would help here, and before and after the transpose below as well. See the function `LSTM_RNN` as an example here: 
https://github.com/guillaume-chevalier/LSTM-Human-Activity-Recognition

Reply:
db2630a